covid_moonshot/analysis/structures.py

```diff
@@ -0,0 +1,316 @@
+    pdbfile_path = f"{project_path}/RUNS/RUN{run}/hybrid_complex.pdb"
+
+    # TODO: Reuse path logic from covid_moonshot.lib
+    trajectory_path = (
+        f"{project_data_path}/RUN{run}/CLONE{clone}/results{gen}/positions.xtc"
+    )
+    pdbfile = md.load(pdbfile_path)
+    trajectory = md.load(trajectory_path, top=pdbfile.top)
+
+    return trajectory
+
+
+def load_fragment(fragment_id: str) -> md.Trajectory:
+    """
+    Load the reference fragment structure
+
+    Parameters
+    ----------
+    fragment_id : str
+      Fragment ID (e.g. 'x10789')
+
+    Returns
+    -------
+    fragment : mdtraj.Trajectory
+      The fragment structure
+
+    """
+    import mdtraj as md
+
+    # TODO: Put this in the covid-moonshot path, or generalize to an arbitrary file
+    fragment = md.load(f"/home/server/Mpro-{fragment_id}_0_bound-protein-thiolate.pdb")
+
+    return fragment
+
+
+def mdtraj_to_oemol(snapshot: md.Trajectory):
+    """
+    Create an OEMol from an MDTraj file by writing and reading
+
+    NOTE: This uses terrible heuristics
+
+    Parameters
+    ----------
+    snapshot : mdtraj.Trajectory
+       MDTraj Trajectory with a single snapshot
+
+    Returns
+    -------
+    oemol : openeye.oechem.OEMol
+       The OEMol
+
+    """
+    from openeye import oechem
+    import tempfile
+    import os
+
+    with tempfile.TemporaryDirectory() as tmpdir:
+        filename = os.path.join(tmpdir, "tmp.pdb")
+        # Write the PDB file
+        snapshot.save(filename)
+        # Read it with OpenEye
+        with oechem.oemolistream(filename) as ifs:
+            for mol in ifs.GetOEGraphMols():
+                return mol
+
+
+def extract_snapshot(
+    project_path: str,
+    project_data_path: str,
+    run: int,
+    clone: int,
+    gen: int,
+    frame: int,
+    fragment_id: str = "x10789",
+):
+    """
+    Extract the specified snapshot, align it to the reference fragment, and write protein and ligands to separate PDB files
+
+    Parameters
+    ----------
+    project_path : str
+       Path to project directory (e.g. '/home/server/server2/projects/13422')
+    run : str or int
+       Run (e.g. '0')
+    clone : str or int
+       Clone (e.g. '0')
+    gen : str or int
+       Gen (e.g. '0')
+    fragment_id : str
+      Fragment ID (e.g. 'x10789')
+
+    Returns
+    -------
+    sliced_snapshot : dict of str : mdtraj.Trajectory
+      sliced_snapshot[name] is the Trajectory for name in ['protein', 'old_ligand', 'new_ligand', 'old_complex', 'new_complex']
+    components : dict of str : oechem.OEMol
+      components[name] is the OEMol for name in ['protein', 'old_ligand', 'new_ligand']
+
+    """
+    # Load the trajectory
+    trajectory = load_trajectory(project_path, project_data_path, run, clone, gen)
+
+    # Load the fragment
+    fragment = load_fragment(fragment_id)
+
+    # Align the trajectory to the fragment (in place)
+    trajectory.superpose(fragment, atom_indices=fragment.top.select("name CA"))
```

Comment:
@mcwitt could you please make sure the line

`trajectory.image_molecules(inplace=True)`

is put in above the `superpose` call please!

Reply:
Added in 3d81f6e